sia/core.go

```diff
@@ -36,31 +55,70 @@ type Core struct {
 // createCore creates a server, host, miner, renter and wallet and
 // puts it all in a single environment struct that's used as the state for the
 // main package.
-//
-// TODO: swap out the way that CreateCore is called so that the wallet,
-// host, etc. can all be used as input - or not supplied at all.
-func CreateCore(hostDir string, walletFile string, serverAddr string, nobootstrap bool) (c *Core, err error) {
+func CreateCore(config Config) (c *Core, err error) {
+	if config.Host == nil {
+		err = errors.New("cannot have nil host")
+		return
+	}
+	if config.HostDB == nil {
+		err = errors.New("cannot have nil hostdb")
+		return
+	}
+	if config.Miner == nil {
+		err = errors.New("cannot have nil miner")
+		return
+	}
+	if config.Renter == nil {
+		err = errors.New("cannot have nil renter")
+		return
+	}
+	if config.Wallet == nil {
+		err = errors.New("cannot have nil wallet")
+		return
+	}
+
+	// Set the number of procs equal to the number of cpus.
+	runtime.GOMAXPROCS(runtime.NumCPU())
+
+	// Fill out the basic information.
 	c = &Core{
-		friends:         make(map[string]consensus.CoinAddress),
+		host:   config.Host,
+		hostDB: config.HostDB,
+		miner:  config.Miner,
+		renter: config.Renter,
+		wallet: config.Wallet,
+
+		// friends:         make(map[string]consensus.CoinAddress),
+
 		blockChan:       make(chan consensus.Block, 100),
 		transactionChan: make(chan consensus.Transaction, 100),
-		hostDir:         hostDir,
-		walletFile:      walletFile,
+
+		hostDir:    config.HostDir,
+		walletFile: config.WalletFile,
 	}
+
+	// Create a state.
 	var genesisOutputDiffs []consensus.OutputDiff
 	c.state, genesisOutputDiffs = consensus.CreateGenesisState()
-	// c.hostDatabase = CreateHostDatabase()
-	// c.host = CreateHost()
-	c.hostDB = hostdb.New()
-	c.miner = miner.New(c.blockChan, 1)
-	// c.renter = CreateRenter()
-	c.wallet, err = wallet.New(c.walletFile)
+	genesisBlock, err := c.state.BlockAtHeight(0)
 	if err != nil {
 		return
 	}
 
 	// Update componenets to see genesis block.
-	err = c.updateMiner(c.miner)
+	err = c.UpdateHost(components.HostAnnouncement{})
+	if err != nil {
+		return
+	}
+	err = c.hostDB.Update(0, nil, []consensus.Block{genesisBlock})
+	if err != nil {
+		return
+	}
+	err = c.UpdateMiner(c.miner.Threads())
+	if err != nil {
+		return
+	}
+	err = c.UpdateRenter()
```

Comment:
these `Update` functions are quite opaque. `hostDB.Update` could have a more descriptive name, and I have no idea what `UpdateMiner` or `UpdateRenter` are doing.